test/e2e/user_quota_tracing/user_quota_tracing_test.go

```diff
@@ -0,0 +1,261 @@
+		}{
+			{"teacher-app-01", queuePath[1], users[0], []string{groups[0]}},
+			{"students-app-01", queuePath[2], users[1], []string{groups[1]}},
+			{"assistant-app-01", queuePath[2], users[2], []string{groups[1]}},
+			{"assistant-app-02", queuePath[1], users[2], []string{groups[0]}},
+		}
+		for _, config := range configs {
+			By(fmt.Sprintf("Deploy the sleep app %s to the %s namespace", config.AppID, ns))
+			sleepObj, podErr := k8s.InitSleepPod(k8s.SleepPodConfig{
+				AppID:  config.AppID,
+				NS:     ns,
+				Labels: map[string]string{"queue": config.Queue},
+			})
+			sleepObj.ObjectMeta.Annotations["yunikorn.apache.org/user.info"] = fmt.Sprintf("{username:%s, groups:{%s}}", config.UserName, config.GroupsNames)
+			Ω(podErr).NotTo(HaveOccurred())
+			sleepRespPod, podErr := kClient.CreatePod(sleepObj, ns)
+			Ω(podErr).NotTo(HaveOccurred())
+
+			// Wait for pod to move to running state
+			podErr = kClient.WaitForPodBySelectorRunning(ns,
+				fmt.Sprintf("app=%s", sleepRespPod.ObjectMeta.Labels["app"]), WAIT_INTERVAL)
+			Ω(podErr).NotTo(HaveOccurred())
+		}
+
+		type Queue struct {
+			RunningApplications []string
+			Resources           map[string]int64
+		}
+		type GroupQuota struct {
+			RunningApplications []string
+			Queues              map[string]Queue
+		}
+		type GroupsQuota struct {
+			NumberOfGroups int
+			Groups         map[string]GroupQuota
+		}
+		expectedGroups := GroupsQuota{
+			2,
+			map[string]GroupQuota{
+				groups[0]: {
+					[]string{configs[0].AppID, configs[3].AppID},
+					map[string]Queue{
+						queuePath[0]: {[]string{configs[0].AppID, configs[3].AppID}, twoPods},
+						queuePath[1]: {[]string{configs[0].AppID, configs[3].AppID}, twoPods},
+						queuePath[2]: {[]string{}, zeroPod},
+					},
+				},
+				groups[1]: {
+					[]string{configs[1].AppID, configs[2].AppID},
+					map[string]Queue{
+						queuePath[0]: {[]string{configs[1].AppID, configs[2].AppID}, twoPods},
+						queuePath[1]: {[]string{}, zeroPod},
+						queuePath[2]: {[]string{configs[1].AppID, configs[2].AppID}, twoPods},
+					},
+				},
+			},
+		}
+		restClient := yunikorn.RClient{}
+		var usedResource yunikorn.ResourceUsage
+		for group, quota := range expectedGroups.Groups {
+			By(fmt.Sprintf("GroupTracker: Check group resource usage of %s in each queue", group))
+			groupUsage, getGroupErr := restClient.GetGroupResourceUsage(DEFAULT_PARTITION, group)
+			Ω(getGroupErr).NotTo(HaveOccurred())
+			Ω(groupUsage.Applications).To(Equal(quota.RunningApplications), "running application IDs are not expected")
+			for queuePath, queueQuota := range quota.Queues {
+				queue, getQueueErr := yunikorn.GetQueueResourceUsage(groupUsage.Queues, queuePath)
+				Ω(getQueueErr).NotTo(HaveOccurred())
+				Ω(queue.RunningApplications).To(Equal(queueQuota.RunningApplications))
+				usedResource.ParseResourceUsage(yunikorn.ParseResource(queue.ResourceUsage))
+				Ω(usedResource).To(Equal(queueQuota.Resources))
+			}
+		}
```

Comment:
I'm wondering if using functions is better... because you're trying to avoid code duplication, which is good.
But perhaps having all these checks in a separate func is more readable, like:

```
type PerGroup struct {
  queuePath string
  runningApps []string
  resources map[string]int64
}

func checkGroups(group string, expected []PerGroup) {
  ... do things with "group" ...
 groupUsage := restClient.GetGroupResourceUsage(DEFAULT_PARTITION, group)
  for _, exp := range expected {
    checkForGroup(groupUsage, exp.queuePath, exp.runningApps, exp.resources)
  }
}

func checkForGroup(groupUsage *dao.GroupResourceUsageDAOInfo, queuePath string, expectedRunningApps []string,
expectedResource map[string]int64) {
 queue, getQueueErr := yunikorn.GetQueueResourceUsage(groupUsage.Queues, queuePath)
 Ω(getQueueErr).NotTo(HaveOccurred())
 Ω(queue.RunningApplications).To(Equal(expectedRunningApps))
 usedResource.ParseResourceUsage(yunikorn.ParseResource(queue.ResourceUsage))
 Ω(usedResource).To(Equal(expectedResource))
}
```

Then you call `checkGroups()` separately for each group, and you create `[]PerGroup` inline with string literals:
```
checkGroups("group1", []PerGroup{
  {
    queuePath: "root.path1",
    runningApps: "app1", "app2", "app3",
    resources: map[string]int64{
      "cpu": 1,
      "memory": 100,
   },
  {
    queuePath: "root.path2",
    runningApps: "app4", "app5",
    resources: map[string]int64{
      "cpu": 2,
      "memory": 200,
   },
  }
})
```
I'm not saying it has to look like this (it might not even compile), but it's a start.

The main idea here is to reduce nesting and the number of structs. I'm really having a hard time understanding what the code does with all these types.

Reply:
Hi @pbacsko , thanks for review.
Your suggestion sounds great in order to reduce the code and make code readable.
I  will try to adopt this suggestion to make this code better.
I am going to start the progress in this weekend.